Calling similar(...) on AbstractCuSparseArray with dims

birkmichael opened the pull request:
Calling `similar(var,dims)` or `similar(var,type,dims)` on a variable `var ` with `typeof(var):<AbstractCuSparseArray` will return a `CuArray` of eltype `eltype(var)` with dimensions given by `dims`. Previously, calling `similar(reshape(var,dims))` would result in Base.abstractarray.jl and Base.reshapedarray.jl expanding the call to `similar(reshape(var,dims),eltype,dims)`, which was not handled in this code and resulted in a call to `Array(...)` and not `CuArray(...)`, forcing use of CPU and scalar indexing.

birkmichael committed:
calling similar on AbstractCuSparseArray with dims

Calling similar() on a variable var of type :<AbstractCuSparseArray along with dims argument will return a CuArray of type eltype(var) with dimensions given by dims.

birkmichael committed:
Merge pull request #1 from birkmichael/CuSparse_Similar

calling similar on AbstractCuSparseArray with dims

maleadt commented:
Thanks for the contribution! I'm not sure I fully understand the fix though, shouldn't `similar` on a sparse array yield another sparse array, not a CuArray?

```
julia> similar(sprand(2,2,0.5))
2×2 SparseMatrixCSC{Float64, Int64} with 3 stored entries:
 6.90942e-310  6.90965e-310
 6.90941e-310   ⋅ 

julia> similar(cu(sprand(2,2,0.5)))
2×2 CuSparseMatrixCSC{Float32, Int32} with 2 stored entries:
  ⋅    ⋅ 
 0.0  1.875
```

Furthermore, a wrapped array isn't a subtype of the original array:

```
julia> reshape(cu(sprand(2,2,0.2)), 2, 2, 1) isa CUSPARSE.AbstractCuSparseArray
false
```

I've also marked your PR as draft so that it doesn't kick off unnecessary CI. It would need some tests for that to be useful anyway.

maleadt commented on lib/cusparse/array.jl:
```diff
@@ -200,6 +200,10 @@ Base.similar(Mat::CuSparseMatrixCSC) = CuSparseMatrixCSC(copy(Mat.colPtr), copy(
 Base.similar(Mat::CuSparseMatrixCSR) = CuSparseMatrixCSR(copy(Mat.rowPtr), copy(Mat.colVal), similar(nonzeros(Mat)), Mat.dims)
 Base.similar(Mat::CuSparseMatrixBSR) = CuSparseMatrixBSR(copy(Mat.rowPtr), copy(Mat.colVal), similar(nonzeros(Mat)), Mat.blockDim, Mat.dir, nnz(Mat), Mat.dims)
 
+Base.similar(a::AbstractCuSparseArray{Tv, <:Any, <:Any}, dims::Base.Dims{N}) where {Tv,N} =
```
You can drop the trailing `<:Any`s in definitions like that.

birkmichael replied:
What should I write down in the curly brackets in the second function, where all type arguments are `<:Any`?
just `a::AbstractCuSparseArray`, dropping the brackets?

maleadt replied:
Yes, and you can use `AbstractCuSparseArray{Tv}` for the first.

birkmichael commented:
> Furthermore, a wrapped array isn't a subtype of the original array:

When `similar` is called on a reshaped array (CPU or GPU), `parent` is called to unwrap, and then `similar` is called on the unwrapped array. This is good behavior for GPU arrays as well. The problem is that when the call is made on the unwrapped array, `eltype` and `dims` are added as arguments, which is currently not handled by CuSparse, so that `similar` is called from `Base` and generates an `Array` rather than a `CuArray`.

> shouldn't similar on a sparse array yield another sparse array, not a CuArray?

You're right. I thought I was mimicing julia's regular sparse behavior, but I just realized how and why `similar(CPUSparseVar,dims::Dims{N>=3})` returns a CPUArray. I will fix my branch and resubmit.

birkmichael committed:
robust handling of similar CSC/CSR

Added more robust handling of applying similar on a CuSparseMatrixCSC/CSR, following the logic used in SparseArrays.jl

birkmichael committed:
Merge pull request #2 from birkmichael/CuSparse_Similar

robust handling of similar CSC/CSR

birkmichael committed:
fixed getrowptr,colvals

two methods CSR were originally added as extensions to SparseArrays.jl, but SparseArrays.jl does not support CSR.

birkmichael committed:
Merge pull request #3 from birkmichael/CuSparse_Similar

fixed getrowptr,colvals

birkmichael committed:
forgot Base. prefix to several instances of similar

birkmichael committed:
Merge pull request #4 from birkmichael/CuSparse_Similar

forgot Base. prefix to several instances of similar

birkmichael committed:
used fill instead of CUDA.fill in matrix initialization

birkmichael committed:
Merge pull request #5 from birkmichael/CuSparse_Similar

used fill instead of CUDA.fill in matrix initialization

birkmichael committed:
tests for similar, unfinished

birkmichael committed:
Created test set for 'similar'

birkmichael committed:
imported some SparseArrays methods

birkmichael committed:
Fixed CSR rowPtr setting in 'similar'

birkmichael committed:
tests for 'similar' for CSC,CSR

unclear what is supported in CUDA.jl for BSR,COO, so they remain untested.

birkmichael committed:
Merge pull request #6 from birkmichael/CuSparse_Similar_tests

Cu sparse Base.similar tests completed for CSC and CSR. No tests for BSR, COO, as it is unclear to what extent these are supported in CUDA.jl.

birkmichael committed:
splatted 1D tuple dims when creating vector

birkmichael committed:
Merge pull request #7 from birkmichael/vectorfix

splatted 1D tuple dims when creating vector

birkmichael commented:
@maleadt So it looks like the tests failed, but not because of me. Some error in the comparison `Array(dx[:])==x[:]` in tests/cusparse.jl.

maleadt commented:
> So it looks like the tests failed, but not because of me. Some error in the comparison `Array(dx[:])==x[:]` in tests/cusparse.jl.

CI on master is green, the failure is in `tests/cusparse` and this PR is changing stuff with sparse arrays, so...
What makes you say the failure is not related?

birkmichael commented:
Two reasons: 
1) I added cusparse/tests/array.jl but didn't touch tests/cusparse.jl
2) the tests fail at the conversion between some cuArray to an Array, and I don't think the `similar` function is called in that process. I will take a closer look.

birkmichael commented:
I found where my changes come up, but I'm not sure how exactly they cause the bug.

- In the test, a sparse matrix `x` is created, and then `d_x` is created via `d_x=CuSparseMatrixCSC(x)`. The test is `Array(d_x[:])==x[:]`. 
- The error is caused by running `d_x[:]`. In master, this creates a normal CPU array (which is wrong, but works out for this test in particular), while in this branch it bugs out. 
- The base of the behaviour seems to be that `_unsafe_getindex(...)@ Base ./multidimensional.jl:845` calls `similar` on `d_x` reshaped to be a vector, a line before the exception is thrown - in master, this just returns a cpu array - this is basically issue #1179. In this branch, it returns a CuSparseVector of the right dimension. 
There's something strange in the way the indices are passes around that makes julia compare (Base.OneTo(875),) and (875, 1), but I can't figure it out.

The stacktrace is:

```
cusparse: Error During Test at /var/lib/buildkite-agent/builds/rtx2080-hydor-elis-ugent-be/julialang/cuda-dot-jl/test/cusparse.jl:44
--
  | Test threw exception
  | Expression: Array(d_x[:]) == x[:]
  | DimensionMismatch("output array is the wrong size; expected (Base.OneTo(875),), got (875, 1)")
  | Stacktrace:
  | [1] throw_checksize_error(A::CuSparseVector{Float64, Int32}, sz::Tuple{Base.OneTo{Int64}})
  | @ Base ./multidimensional.jl:881
  | [2] _unsafe_getindex(#unused#::IndexCartesian, A::Base.ReshapedArray{Float64, 1, CuSparseMatrixCSC{Float64, Int32}, Tuple{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64}}}, I::Base.Slice{Base.OneTo{Int64}})
  | @ Base ./multidimensional.jl:845
  | [3] _getindex
  | @ ./multidimensional.jl:832 [inlined]
  | [4] getindex(A::CuSparseMatrixCSC{Float64, Int32}, I::Function)
  | @ Base ./abstractarray.jl:1170
  | [5] macro expansion
  | @ /var/lib/buildkite-agent/builds/rtx2080-hydor-elis-ugent-be/julialang/cuda-dot-jl/test/cusparse.jl:44 [inlined]
  | [6] (::var"#5#8")()
  | @ Main ~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/packages/GPUArrays/0vqbc/src/host/indexing.jl:75
  | [7] task_local_storage(body::var"#5#8", key::Symbol, val::Bool)
  | @ Base ./task.jl:281
  | [8] macro expansion
  | @ ~/.cache/julia-buildkite-plugin/depots/3cc01fab-3357-4a7a-9294-cde2d3115a97/packages/GPUArrays/0vqbc/src/host/indexing.jl:74 [inlined]
  | [9] macro expansion
  | @ /var/lib/buildkite-agent/builds/rtx2080-hydor-elis-ugent-be/julialang/cuda-dot-jl/test/cusparse.jl:43 [inlined]
  | [10] macro expansion
  | @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
  | [11] top-level scope
  | @ /var/lib/buildkite-agent/builds/rtx2080-hydor-elis-ugent-be/julialang/cuda-dot-jl/test/cusparse.jl:15
```

birkmichael committed:
Merge pull request #8 from birkmichael/CuSparse_Similar

Cu sparse similar

birkmichael committed:
Added method getindex(CuSparseMat,Colon), changed show for vectors

birkmichael committed:
Added conversion from matrix to vector

birkmichael committed:
Merge pull request #9 from birkmichael/CuSparse_similar_toVector

Added method for colon indexing a CuSparseMatrix, and a conversion from CuSparseMatrix to CuSparseMatrix to vector. Fixed show method of vectors to be same as for SparseArrays. CuSparseVector.dims is a 2-tuple instead of an integer, probably as a holdover from the base.show method that was written to fit both vectors and matrices, and needed to be able to run 'size(A)[2]'. That can probably be changed now, but I don't want to break anyone's code.

birkmichael committed:
Merge branch 'master' into CuSparse_Similar

birkmichael commented:
@maleadt  It looks like there were a bunch of tests that did indexing of a CuSparseMatrix and compared them to a CPU sparse matrix, a la `Array(d_x[inds])==x[inds]`. All of these worked in master because indexing with just a single colon or with ranges/vectors just turned the CuSparseMatrix into a dense CPU array, and then indexed into that (even before `Array()` was called). Now that `similar` is fixed to return the correct  type of array, the lack of specialized methods for indexing CuSparseMatrix variables is causing errors. I'm thinking of solving all of these bugs just by converting to a sparse CPU array, letting SparseArrays do its magic, and then convert back to a CuSparseMatrix. It looks like these operations demand scalar indexing anyways, so it's not much of a performance loss. What do you think?

maleadt commented:
Copying an entire sparse array to the CPU when you only want to access a single element seems excessive... Scalar iteration is slow because it performs a `memcpy` for every element, that doesn't mean we should make it even slower by `memcpy`ing the entire array for each access.

birkmichael commented:
It's not for a single element. Specifically, what exists now (and returns the right type):

1. indexing with a single integer and a colon (both arrangements - row slicing and column slicing)
2. indexing with two colons
3. indexing with two integers
4. indexing with a single colon - added in my PR, converts to a CPU sparse array, uses their algorithm to convert to a sparse vector (with scalar iteration), and then converts to a CuSparseVector.

What I'm talking about is indexing where one or both of the indices is a vector or range. In master, if one is a vector/range, you will get a cpu array. If both are, you will get an error (which is why the test failed). The simplest fix is by converting to CPU and back. Admittedly, this is probably not the most efficient way to go about adding those features, but I'm not sure there is a good method to do this kind of slicing without doing scalar iteration anyway, which is why I am proposing this method.

birkmichael committed:
Full indexing of CuSparseArrays

birkmichael committed:
Vectorising matrices

birkmichael committed:
Merge branch 'JuliaGPU:master' into master

birkmichael committed:
Added vectorising of CuSparseMatrices

birkmichael committed:
Update array.jl

birkmichael committed:
Update conversions.jl

birkmichael committed:
Update conversions.jl

birkmichael committed:
Merge pull request #11 from birkmichael/CuSparse_Similar_matrixIndexing

Cu sparse similar matrix indexing

birkmichael committed:
changed to match updates in master

birkmichael committed:
Update conversions.jl

birkmichael committed:
Merge pull request #12 from JuliaGPU/master

updating to match current master branch

birkmichael committed:
Merge branch 'CuSparse_Similar' into master

birkmichael committed:
Merge pull request #13 from birkmichael/master

Similar, indexing CuSparseArrays, Vector reshaping

birkmichael commented:
All tests passed except those that fail `master`. Added a bunch of new tests (tests/cusparse/arrays.jl), so the coverage will probably improve. Anyway, now you can index and slice CuCSC and CuCSR matrices freely and they will stay the same type, rather than drop to the CPU. Any type of AbstractCuSparseMatrix can be turned into a CuSparseVector now. Everything is done using (almost only) vectorised methods without going through the CPU.

maleadt commented on lib/cusparse/array.jl:
```diff
@@ -200,6 +200,59 @@ Base.similar(Mat::CuSparseMatrixCSC) = CuSparseMatrixCSC(copy(Mat.colPtr), copy(
 Base.similar(Mat::CuSparseMatrixCSR) = CuSparseMatrixCSR(copy(Mat.rowPtr), copy(Mat.colVal), similar(nonzeros(Mat)), Mat.dims)
 Base.similar(Mat::CuSparseMatrixBSR) = CuSparseMatrixBSR(copy(Mat.rowPtr), copy(Mat.colVal), similar(nonzeros(Mat)), Mat.blockDim, Mat.dir, nnz(Mat), Mat.dims)
 
+## similar for CSC,CSR with dims, eltype arguments, adapted from SparseArrays.jl. NOTE: calling similar() on a wrapped CuSparseMatrixBSR/COO will result in a CuArray.
```
Maybe wrap the `NOTE` on a new line.

But also: why does `similar` still result in a dense vector? https://github.com/JuliaGPU/CUDA.jl/pull/1184#issuecomment-933626320

birkmichael replied:
It shouldn't - I spent a lot of time on that. I'll check.

maleadt commented on lib/cusparse/array.jl:
```diff
@@ -200,6 +200,59 @@ Base.similar(Mat::CuSparseMatrixCSC) = CuSparseMatrixCSC(copy(Mat.colPtr), copy(
+
+# Handles cases where CSC/CSR are reshaped to more than two dimensions, and all cases above for COO/BSR.
+Base.similar(a::AbstractCuSparseArray, ::Type{T}, dims::Base.Dims{N}) where {T,N} =
+CuArray{T,N}(undef, dims)               
```
indent?

Comment from above also applies.

maleadt commented on lib/cusparse/array.jl:
```diff
@@ -260,29 +316,62 @@ Hermitian{T}(Mat::CuSparseMatrix{T}) where T = Hermitian{T,typeof(Mat)}(Mat,'U')
-    r1 = convert(Int, x.colPtr[j])
-    r2 = convert(Int, x.colPtr[j+1]) - 1
-    CuSparseVector(rowvals(x)[r1:r2], nonzeros(x)[r1:r2], size(x, 1))
+Base.getindex(A::CuSparseMatrixCSC,I::AbstractVector,J::AbstractVector) = CuSparseMatrixCSC(getindex(CuSparseMatrixCSR(getindex(A,:,J)),I,:))
```
Maybe wrap these two methods for legibility.

birkmichael replied:
How do you mean? Write a method for `Base.getindex(A::CuSparseMatrixCSCR,I::AbstractVector,J::AbstractVector)` that calls either of these two methods, depending on type?

maleadt replied:
No, just line wrapping :-) i.e. adding a newline after the `=`

maleadt commented on lib/cusparse/array.jl:
```diff
@@ -260,29 +316,62 @@ Hermitian{T}(Mat::CuSparseMatrix{T}) where T = Hermitian{T,typeof(Mat)}(Mat,'U')
+    newcolptr = cumsum(diffVec)
+
+    indices = map((x,y)->range(x,stop=y),colptr[I_sorted],colptr[I_sorted.+1].-1)
+    @CUDA.allowscalar indices = vcat(indices...) #TODO: figure out how to avoid this step
```
`CUDA` module prefix shouldn't be needed.

But the scalar indexing is not good. where does it come from?

birkmichael replied:
What I'm trying to do here is this: get an array `indices` of vectors, where `indices[i]` contains a vector of the indices of all elements in `nzvals` that are in the `I_sorted[i]` row (or column, depending on CSC or CSR) of the input matrix. After that, indexing becomes simple.

However, the `map` here returns a CuArray of ranges, and any indexing I do with that will cause a bug, so I need to turn the ranges into vectors first. If I had a method to get a CPU array of CuVectors from the `map`, instead of a `CuArray` of ranges, that would be unnecessary. Applying `collect.(indices)` gives me an "inline variables only" type of error because it tries to instantiate a CuArray of arrays of (sometimes) different sizes. Applying `vcat(indices...)` gives a scalar indexing error.

The other, more elegant method of solving this, is actually doing what I did in `conversions.jl` that you commented on - transforming `colptr` to a CPU array, in which case I can just get an array of arrays using a comprehension or a map.

maleadt replied:
> the map here returns a CuArray of ranges, and any indexing I do with that will cause a bug, so I need to turn the ranges into vectors first.

Could you elaborate? And conversions to CPU data is extremely expensive and to be avoided if possible, so I'd rather look into fixing the bug you're encountering.

maleadt commented on lib/cusparse/array.jl:
```diff
@@ -446,6 +535,7 @@ Base.copy(Mat::CuSparseMatrixCOO) = copyto!(similar(Mat), Mat)
 
 # input/output
 
+
```
Some needless whitespace changes?

maleadt commented on lib/cusparse/array.jl:
```diff
@@ -483,7 +573,6 @@ for (gpu, cpu) in [CuSparseMatrixCSC => SparseMatrixCSC,
     end
 end
 
-
```
Some needless whitespace changes?

maleadt commented on test/cusparse/array.jl:
```diff
@@ -0,0 +1,143 @@
+using CUDA.CUSPARSE
+using SparseArrays
+using Test
```
You shouldn't need to import `Test` -- run select tests by doing `julia -L test/setup.jl test/cusparse/array.jl`

maleadt commented on lib/cusparse/conversions.jl:
```diff
@@ -41,6 +41,24 @@ function SparseArrays.sparse(I::CuVector{Cint}, J::CuVector{Cint}, V::CuVector{T
+    I = 1:m
+    nzval = nonzeros(Mat)
+    nzind = rowvals(Mat)
+    colptr = Array(getcolptr(Mat)) #TODO: figure out how to avoid this step
```
this seems bad. there's no API call for such conversions? or maybe try doing the conversion on the GPU to avoid synchronizing during the copy to/from the CPU?

birkmichael replied:
See my comment Re: scalar indexing above - there I try doing it on the GPU and end up doing scalar indexing. Do you mean API calls to the NVidia CUDA library? I had a look now and couldn't find it, but I will give it a bit more effort later.

maleadt reviewed (commented):
Thanks! I don't have the time for an in depth review right now, so some quick comments.
